v-next/hardhat/src/internal/builtin-plugins/run/index.ts

```diff
@@ -0,0 +1,23 @@
+import type { HardhatPlugin } from "@nomicfoundation/hardhat-core/types/plugins";
+
+import { ParameterType, task } from "@nomicfoundation/hardhat-core/config";
+
+import { runScriptWithHardhat } from "./runScriptWithHardhat.js";
+
+export default {
+  id: "run",
+  tasks: [
+    task("run", "Runs a user-defined script after compiling the project")
+      .addPositionalParameter({
+        name: "script",
+        description: "A js or ts file to be run within hardhat's environment",
+        type: ParameterType.STRING,
+      })
+      .addFlag({
+        name: "noCompile",
+        description: "Don't compile before running this task",
+      })
+      .setAction(runScriptWithHardhat)
```

Comment:
I think we should do this instead

```suggestion
      .setAction(import.meta.resolve('./runScriptWithHardhat.js'))
```

and export the function as default within that file.

This way the implementation of the task is only loaded if it's run.